### PR TITLE
Added tests for logging using mock

### PR DESCRIPTION
**Pull request by sreecharansankaranarayanan**

### CODE REVIEW

File: notifications/test_pipeline.py
Comment (jcdyer) [inline notifications/test_pipeline.py:34]: Looks pretty good, but you should add another similar test that shows what happens when a message is expired, and catches some of the included warnings.

Comment (sreecharansankaranarayanan) [reply notifications/test_pipeline.py:34]: Is it enough to simply print out the logging info and logging warning messages (if they exist)?

Comment (jcdyer) [reply notifications/test_pipeline.py:34]: I think what you're doing here (checking counts of messages) is sufficient, but we need a test that shows what happens when a message expires on its way through the queue.

**Comment by jcdyer** [comment]
Looks good to me.  Next steps would be:
1. resolving the quality issues.
2. getting a review from @nasthagiri

File: notifications/message.py
Comment (nasthagiri) [inline notifications/message.py]: Need to add `expiration_time` to the docstring.

Comment (nasthagiri) [inline notifications/message.py]: Ending paren should be on its own line.

File: notifications/pipeline_manager.py
Comment (nasthagiri) [inline notifications/pipeline_manager.py]: Nit: It would be better to raise an exception here, instead of having the reader try to figure out whether the value `1` is good or bad.

File: notifications/pipeline_manager.py
Comment (nasthagiri) [inline notifications/pipeline_manager.py]: Nit: `Completed` -> `Aborted` or `Expired`.

File: notifications/test_pipeline.py
Comment (nasthagiri) [inline notifications/test_pipeline.py]: Ending paren should be 4 spaces to the left.  Not sure why it wasn't caught by pylint.  @jcdyer ?

File: notifications/pipeline_steps/base.py
Comment (nasthagiri) [inline notifications/pipeline_steps/base.py]: Nit: We want this method to be implemented by each Pipeline step, correct?  If so, I would want the expiration check to be handled by the `Pipeline.process` method instead - so it is implemented and checked in one place for all pipeline steps.  If you make this change, you'll need to update your test as well.

File: notifications/test_pipeline.py
Comment (sreecharansankaranarayanan) [reply notifications/test_pipeline.py]: @nasthagiri I had it that way but pylint complained for some reason.

File: notifications/message.py
Comment (sreecharansankaranarayanan) [reply notifications/message.py]: Done.

Comment (sreecharansankaranarayanan) [reply notifications/message.py]: Done.

File: notifications/test_pipeline.py
Comment (sreecharansankaranarayanan) [reply notifications/test_pipeline.py]: Fixed it now.

File: notifications/pipeline_manager.py
Comment (sreecharansankaranarayanan) [reply notifications/pipeline_manager.py]: The logging level is set such that the logging warning gets printed on the console. Isn't that sufficient?

File: notifications/pipeline_manager.py
Comment (sreecharansankaranarayanan) [reply notifications/pipeline_manager.py]: Done.

File: notifications/pipeline_steps/base.py
Comment (sreecharansankaranarayanan) [reply notifications/pipeline_steps/base.py]: Checking for expiry at each pipeline step instead allows for dropping of the message as soon as it expires and return to the caller. If we checked before and after the call though, we wouldn't be able to drop a message that was stuck in a step for whatever reason. I think it's desirable to keep it this way.

Comment (nasthagiri) [reply notifications/pipeline_steps/base.py]: The expiration time check would happen microseconds before calling the pipeline step's `process_message` method - since it would be called from within the `Pipeline.process` method _after_ the message is received by the queue/process.  So there is no lag in time that we need to worry about.

File: notifications/pipeline_manager.py
Comment (nasthagiri) [reply notifications/pipeline_manager.py]: I'm not sure I understand what you mean.  I am referring to the code checking whether the return value is equal to `1`.  As a reader of this code, how am I to understand the semantics of `1`?
By changing this to an exception, you can treat message expiration as an exceptional case, which would also result in cleaner code.

File: notifications/pipeline_steps/base.py
Comment (sreecharansankaranarayanan) [reply notifications/pipeline_steps/base.py]: As it stands, that's true but in the future, each of the processing steps might want to drop the message if it expires at the step itself right? Rather than wait for the processing to complete return and then drop the message.

File: notifications/pipeline_manager.py
Comment (sreecharansankaranarayanan) [reply notifications/pipeline_manager.py]: I get it. I thought you were talking about writing to console.

File: notifications/pipeline_steps/base.py
Comment (nasthagiri) [reply notifications/pipeline_steps/base.py]: If the check happens at L65 in pipeline_manager.py right before calling process_message, the check will happen  right after the message is dequeued and immediately before calling the specific pipeline step's `process_message`.  Once again, we don't want each step to implement its own expiration check.  That should be verified within a single place.

File: notifications/message.py
Comment (nasthagiri) [inline notifications/message.py]: Why is time localization needed here?  The `utcnow` call will return a UTC timestamp, which should be used for all time computations and comparisons at this layer.

File: notifications/pipeline_manager.py
Comment (nasthagiri) [inline notifications/pipeline_manager.py]: See above comment about not needing to localize the times.  This shouldn't be needed at this layer.

File: notifications/pipeline_manager.py
Comment (nasthagiri) [inline notifications/pipeline_manager.py]: Nit: This can be DRYed - These exact lines are repeated 2x.

File: notifications/test_pipeline.py
Comment (nasthagiri) [inline notifications/test_pipeline.py]: Commented out code shouldn't be merged.

**Review by nasthagiri — COMMENTED** [review]
A few remaining comments.

File: notifications/message.py
Comment (sreecharansankaranarayanan) [reply notifications/message.py]: That's right. Time localization needs to be done only at the presentation layer.

File: notifications/pipeline_manager.py
Comment (sreecharansankaranarayanan) [reply notifications/pipeline_manager.py]: They differ slightly because they capture dropping before and after processing by the step but I am going to see if it can be improved.

**Review by nasthagiri — APPROVED** [review]
Looks great.  Please remember to squash before your merge.  
Also, if you want extra-credit, consider verifying the contents of the log message and not just the number of log messages.

**Pull request merged**
